manta-trusted-setup/src/ceremony/signature.rs

```diff
@@ -0,0 +1,250 @@
+        CeremonyError,
+    };
+    use alloc::vec::Vec;
+    use ed25519_dalek::{self, Signer, Verifier};
+    use manta_util::{
+        into_array_unchecked,
+        serde::{Deserialize, Serialize},
+        serde_with::serde_as,
+    };
+
+    /// The dalek implementation of `ed25519` signatures.
+    pub struct Ed25519 {}
+
+    /// The public key for signed messages from contributors. This is a wrapper around the
+    /// byte representation of an `ed25519_dalek::PublicKey` type.  The original type does
+    /// not implement `Hash` and so cannot be used as a key in the `Registry`.
+    #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Ord, PartialOrd)]
+    pub struct PublicKey(pub [u8; 32]);
+
+    /// The private key for signed messages from contributors. This is a wrapper around the
+    /// byte representation of an `ed25519_dalek::SecretKey` type. The byte representation
+    /// is used to be consistent with the choice made for `ContributorPublicKey`.
+    #[derive(Debug)]
+    pub struct PrivateKey(pub [u8; 32]);
+
+    /// The type for message signatures.
+    #[serde_as]
+    #[derive(Debug, Serialize, Deserialize, Copy, Clone)]
+    pub struct Signature(
+        #[serde_as(as = "[_; ed25519_dalek::Signature::BYTE_SIZE]")]
+        [u8; ed25519_dalek::Signature::BYTE_SIZE],
+    );
+
+    impl From<ed25519_dalek::Signature> for Signature {
+        fn from(f: ed25519_dalek::Signature) -> Self {
+            Signature(f.to_bytes())
+        }
+    }
+
+    impl From<Signature> for ed25519_dalek::Signature {
+        fn from(f: Signature) -> Self {
+            ed25519_dalek::Signature::from_bytes(&f.0).expect("Should never fail")
+        }
+    }
+
+    impl SignatureScheme for Ed25519 {
+        type PublicKey = PublicKey;
+        type PrivateKey = PrivateKey;
+        type Signature = Signature;
+        type DomainTag = [u8];
+    }
+
+    /// The signable messages
+    pub struct Message<'a>(&'a [u8]);
+
+    impl<'a> From<&'a [u8]> for Message<'a> {
+        fn from(s: &'a [u8]) -> Self {
+            Self(s)
+        }
+    }
+
+    impl<'a> Sign<Ed25519> for Message<'a> {
+        type Signature = Signature;
+
+        fn sign(
+            &self,
+            domain_tag: &[u8],
+            public_key: &<Ed25519 as SignatureScheme>::PublicKey,
+            private_key: &<Ed25519 as SignatureScheme>::PrivateKey,
+        ) -> Result<<Ed25519 as SignatureScheme>::Signature, CeremonyError> {
+            // Read keypair from byte representation of pub/priv keys
+            let keypair = ed25519_dalek::Keypair::from_bytes(
+                &[&private_key.0[..], &public_key.0[..]].concat(),
+            )
+            .expect("Failed to decode keypair from bytes"); // todo: error handling
+
+            let msg = domain_tag
+                .iter()
+                .chain(self.0.iter())
+                .copied()
+                .collect::<Vec<_>>();
+
+            let sig = keypair.sign(&msg);
```

Comment:
It seems that we only sign `domain_tag` here instead of a general message `msg`. Should we explicitly split the concept of `domain_tag` and `msg` here? For example, `domain_tag` should be some `u8` value while `msg` is an array of `u8`.